File: packages/ts-sdk/src/analysis/measures.ts
import {
  positionsToBinary,
  positionsToIntervals,
  PositionVector,
  type BinaryVector
} from "../core";
import {
  type AksakClassification,
  type GeneratedResult,
  type MeasureSummary,
  type OccurrenceCount
} from "./types";
import {
  arraysEqual,
  containsNumber,
  cppPrecision,
  maxNumber,
  minNumber,
  occurrenceCountForDistance
} from "./shared";

export function differences(values: readonly number[]): number[] {
  var result: number[] = [];
  if (values.length < 2) {
    return result;
  }

  for (var index = 1; index < values.length; index += 1) {
    result.push(values[index] - values[index - 1]);
  }

  return result;
}

export function geodesicDistance(left: number, right: number, mod: number): number {
  var distance = ((right - left) + mod) % mod;
  if (distance > (mod / 2)) {
    distance = mod - distance;
  }
  return distance;
}

export function geodesicDistances(value: PositionVector): number[] {
  var result: number[] = [];
  var data = value.toArray();

  for (var leftIndex = 0; leftIndex < data.length; leftIndex += 1) {
    for (var rightIndex = leftIndex + 1; rightIndex < data.length; rightIndex += 1) {
      result.push(geodesicDistance(data[leftIndex], data[rightIndex], value.mod));
    }
  }

  return result;
}

export function isEuclidean(value: PositionVector, mod: number): boolean {
  var intervals = positionsToIntervals(value).toArray();
  var rotated = intervals.slice();
  var baseline = intervals.slice();
  var length = rotated.length;

  if (length === 0) {
    return false;
  }

  rotated[0] += 1;
  rotated[length - 1] -= 1;

  for (var index = 0; index < length; index += 1) {
    if (arraysEqual(rotated, baseline)) {
      return true;
    }
    rotated.push(rotated.shift() as number);
  }

  return false;
}

export function calculateOccurrences(values: readonly number[]): OccurrenceCount[] {
  var counts: Record<string, number> = {};
  var index = 0;

  for (index = 0; index < values.length; index += 1) {
    var key = String(values[index]);
    counts[key] = (counts[key] || 0) + 1;
  }

  var keys = Object.keys(counts).map(function (value) {
    return Number(value);
  }).sort(function (left, right) {
    return left - right;
  });

  var result: OccurrenceCount[] = [];
  for (index = 0; index < keys.length; index += 1) {
    result.push({
      count: counts[String(keys[index])],
      distance: keys[index]
    });
  }

  return result;
}

export function isWinogradDeep(occurrences: readonly OccurrenceCount[], size: number): boolean {
  var seen: number[] = [];

  for (var distance = 1; distance < size; distance += 1) {
    var count = occurrenceCountForDistance(occurrences, distance);
    if (count === null || seen.indexOf(count) !== -1) {
      return false;
    }
    seen.push(count);
  }

  return true;
}

export function isErdosDeep(occurrences: readonly OccurrenceCount[]): boolean {
  var seen: number[] = [];

  for (var index = 0; index < occurrences.length; index += 1) {
    if (seen.indexOf(occurrences[index].count) !== -1) {
      return false;
    }
    seen.push(occurrences[index].count);
  }

  return true;
}

export function calculateRegressionEvenness(rhythm: readonly number[], totalTimeUnits: number): number {
  var numberOfNotes = rhythm.length;
  var idealInterval = totalTimeUnits / numberOfNotes;
  var regressionEvenness = 0;

  for (var index = 0; index < numberOfNotes; index += 1) {
    regressionEvenness += Math.abs(rhythm[index] - (index * idealInterval));
  }

  return regressionEvenness;
}

export function calculateRhythmicOddity(value: PositionVector): number {
  var notes = value.toArray();
  var rhythmicOddity = 0;

  for (var leftIndex = 0; leftIndex < notes.length; leftIndex += 1) {
    for (var rightIndex = leftIndex + 1; rightIndex < notes.length; rightIndex += 1) {
      var firstDistance = ((notes[rightIndex] - notes[leftIndex]) + value.mod) % value.mod;
      var secondDistance = value.mod - firstDistance;
      if (firstDistance === secondDistance) {
        rhythmicOddity += 1;
      }
    }
  }

  return rhythmicOddity;
}

export function computeTransitionComplexity(value: PositionVector, mod: number): number {
  var binary = positionsToBinary(value).toArray();
  var complexity = 0;

  if (binary.length === 0) {
    return complexity;
  }

  for (var index = 1; index < binary.length; index += 1) {
    if (binary[index] !== binary[index - 1]) {
      complexity += 1;
    }
  }

  return complexity;
}

export function computeEntropy(value: PositionVector): number {
  var binary = positionsToBinary(value).toArray();
  var frequencies: Record<string, number> = {};
  var entropy = 0;

  if (binary.length === 0) {
    return 0;
  }

  for (var index = 0; index < binary.length; index += 1) {
    var key = String(binary[index]);
    frequencies[key] = (frequencies[key] || 0) + 1;
  }

  var keys = Object.keys(frequencies);
  for (index = 0; index < keys.length; index += 1) {
    var probability = frequencies[keys[index]] / binary.length;
    entropy -= probability * (Math.log(probability) / Math.log(2));
  }

  return entropy;
}

export function computeLongestSubsequence(value: PositionVector): number {
  var binary = positionsToBinary(value).toArray();
  var longest = 0;
  var currentRun = 0;

  if (binary.length === 0) {
    return 0;
  }

  longest = 1;
  currentRun = 1;

  for (var index = 1; index < binary.length; index += 1) {
    if (binary[index] === binary[index - 1]) {
      currentRun += 1;
      if (currentRun > longest) {
        longest = currentRun;
      }
    } else {
      currentRun = 1;
    }
  }

  return longest;
}

export function calculateDistributionSpectra(value: PositionVector): Array<readonly number[]> {
  var normalizedScale = value.toArray();
  var spectra: number[][] = [];
  var size = normalizedScale.length;
  var leftIndex = 0;
  var rightIndex = 0;

  if (size < 2) {
    return [];
  }

  for (leftIndex = 0; leftIndex < size - 1; leftIndex += 1) {
    spectra.push([]);
  }

  for (leftIndex = 0; leftIndex < size; leftIndex += 1) {
    for (rightIndex = 1; rightIndex < size; rightIndex += 1) {
      var specificInterval = ((normalizedScale[(leftIndex + rightIndex) % size] - normalizedScale[leftIndex]) + value.mod) % value.mod;
      if (!containsNumber(spectra[rightIndex - 1], specificInterval)) {
        spectra[rightIndex - 1].push(specificInterval);
      }
    }
  }

  for (leftIndex = 0; leftIndex < spectra.length; leftIndex += 1) {
    spectra[leftIndex].sort(function (left, right) {
      return left - right;
    });
  }

  return spectra;
}

export function calculateSpectrumWidths(spectra: ReadonlyArray<readonly number[]>): number[] {
  var widths: number[] = [];

  for (var index = 0; index < spectra.length; index += 1) {
    if (spectra[index].length === 0) {
      widths.push(0);
    } else {
      widths.push(maxNumber(spectra[index]) - minNumber(spectra[index]));
    }
  }

  return widths;
}

export function calculateSpectrumVariation(widths: readonly number[], numberOfTones: number): number {
  var sum = 0;

  for (var index = 0; index < widths.length; index += 1) {
    sum += widths[index];
  }

  return sum / numberOfTones;
}

export function findRotationalSymmetryAxes(scale: PositionVector): number[] {
  var normalizedScale = scale.toArray();
  var axes: number[] = [];
  var size = normalizedScale.length;

  for (var interval = 1; interval < scale.mod; interval += 1) {
    var transposedScale: number[] = new Array(size);
    for (var index = 0; index < size; index += 1) {
      transposedScale[index] = (normalizedScale[index] + interval) % scale.mod;
    }
    transposedScale.sort(function (left, right) {
      return left - right;
    });
    if (arraysEqual(transposedScale, normalizedScale)) {
      axes.push(interval);
    }
  }

  return axes;
}

export function findReflectiveSymmetryAxes(scale: PositionVector): number[] {
  var normalizedScale = scale.toArray();
  var axes: number[] = [];
  var size = normalizedScale.length;

  for (var doubledAxis = 0; doubledAxis < (2 * scale.mod); doubledAxis += 1) {
    var axis = doubledAxis / 2;
    var isSymmetric = true;

    for (var index = 0; index < size; index += 1) {
      var reflected = (2 * axis) - normalizedScale[index];
      var reflectedNote = ((reflected + (10 * scale.mod)) % scale.mod);
      if (!containsNumber(normalizedScale, reflectedNote)) {
        isSymmetric = false;
        break;
      }
    }

    if (isSymmetric) {
      axes.push(axis);
    }
  }

  return axes;
}

export function isPrime(value: number): boolean {
  if (value <= 1) {
    return false;
  }
  if (value === 2) {
    return true;
  }
  if ((value % 2) === 0) {
    return false;
  }

  for (var divisor = 3; divisor <= Math.sqrt(value); divisor += 2) {
    if ((value % divisor) === 0) {
      return false;
    }
  }

  return true;
}

export function classifyAksakRhythm(mod: number): AksakClassification {
  if (isPrime(mod)) {
    return "authentic aksak";
  }
  if ((mod % 2) !== 0) {
    return "quasi-aksak";
  }
  return "pseudo-aksak";
}

export function isPalindrome(scale: PositionVector): boolean {
  return findReflectiveSymmetryAxes(scale).indexOf(0) !== -1;
}

export function isChiral(scale: PositionVector): boolean {
  var normalizedScale = scale.toArray();
  var mirroredScale = normalizedScale.slice();
  var size = normalizedScale.length;

  for (var index = 0; index < mirroredScale.length; index += 1) {
    mirroredScale[index] = (scale.mod - mirroredScale[index]) % scale.mod;
  }

  mirroredScale.sort(function (left, right) {
    return left - right;
  });

  if (arraysEqual(normalizedScale, mirroredScale)) {
    return false;
  }

  for (var interval = 1; interval < scale.mod; interval += 1) {
    var transposed: number[] = new Array(size);
    for (index = 0; index < size; index += 1) {
      transposed[index] = (mirroredScale[index] + interval) % scale.mod;
    }
    transposed.sort(function (left, right) {
      return left - right;
    });
    if (arraysEqual(transposed, normalizedScale)) {
      return false;
    }
  }

  return true;
}

export function isBalanced(scale: PositionVector): boolean {
  var xSum = 0;
  var ySum = 0;
  var angleStep = (2 * Math.PI) / scale.mod;
  var data = scale.toArray();

  for (var index = 0; index < data.length; index += 1) {
    var angle = data[index] * angleStep;
    xSum += Math.cos(angle);
    ySum += Math.sin(angle);
  }

  return Math.abs(xSum) < 1e-6 && Math.abs(ySum) < 1e-6;
}

export function generate(m: number, k: number, n: number): number[] {
  var sequence: number[] = [];

  for (var index = 0; index < k; index += 1) {
    sequence.push((index * m) % n);
  }

  return sequence;
}

export function isGenerated(values: readonly number[], n: number): GeneratedResult {
  var size = values.length;

  for (var multiplier = 1; multiplier < n; multiplier += 1) {
    var generated = generate(multiplier, size, n).slice().sort(function (left, right) {
      return left - right;
    });
    if (arraysEqual(generated, values)) {
      return {
        found: true,
        multiplier: multiplier
      };
    }
  }

  return {
    found: false,
    multiplier: -1
  };
}

export function measureSummary(value: PositionVector): MeasureSummary {
  var positions = value.toArray();
  var intervals = positionsToIntervals(value).toArray();
  var onsets = positionsToBinary(value);
  var differencesData = differences(positions);
  var geodesicData = geodesicDistances(value);
  var occurrences = calculateOccurrences(geodesicData);
  var spectra = calculateDistributionSpectra(value);
  var widths = calculateSpectrumWidths(spectra);
  var rotationalAxes = findRotationalSymmetryAxes(value);
  var reflectiveAxes = findReflectiveSymmetryAxes(value);
  var modulo = value.range;
  var generated = isGenerated(positions, modulo);

  return {
    aksakClassification: classifyAksakRhythm(modulo),
    balanced: isBalanced(value),
    chiral: isChiral(value),
    differences: differencesData,
    distributionSpectra: spectra,
    erdosDeep: isErdosDeep(occurrences),
    generated: generated,
    geodesicDistances: geodesicData,
    intervals: intervals,
    isEuclidean: isEuclidean(value, modulo),
    longestSubsequence: computeLongestSubsequence(value),
    modulo: modulo,
    occurrences: occurrences,
    onsets: onsets.toJSON(),
    palindrome: isPalindrome(value),
    positions: value.toJSON(),
    reflectiveSymmetryAxes: reflectiveAxes,
    regressionEvenness: cppPrecision(calculateRegressionEvenness(positions, modulo)),
    rhythmicOddity: calculateRhythmicOddity(value),
    rotationalSymmetryAxes: rotationalAxes,
    spectrumVariation: cppPrecision(calculateSpectrumVariation(widths, value.size)),
    spectrumWidths: widths,
    transitionComplexity: computeTransitionComplexity(value, modulo),
    winogradDeep: isWinogradDeep(occurrences, value.size),
    entropy: cppPrecision(computeEntropy(value))
  };
}

File: packages/dotnet-sdk/src/Measures.cs
namespace Musicpp;

public static class Measures
{
    public static List<int> Differences(IReadOnlyList<int> input)
    {
        if (input.Count < 2)
            return new List<int>();
        var outList = new List<int>();
        for (var i = 1; i < input.Count; i++)
            outList.Add(input[i] - input[i - 1]);
        return outList;
    }

    public static int GeodesicDistance(int a, int b, int mod)
    {
        if (mod <= 0)
            throw new ArgumentException("geodesicDistance: mod must be positive");
        var distance = (b - a + mod) % mod;
        if (distance > mod / 2)
            distance = mod - distance;
        return distance;
    }

    public static List<int> GeodesicDistances(PositionVector input)
    {
        var distances = new List<int>();
        var n = input.Count;
        for (var i = 0; i < n; i++)
        {
            for (var j = i + 1; j < n; j++)
                distances.Add(GeodesicDistance(input[i], input[j], input.Mod));
        }
        return distances;
    }

    public static bool IsEuclidean(PositionVector input, int mod)
    {
        var j = MusicVectors.PositionsToIntervals(input);
        var temp = j.MutableData.ToList();
        var temp2 = temp.ToList();
        var n = temp.Count;
        if (n == 0)
            return false;
        temp[0]++;
        temp[n - 1]--;
        for (var i = 0; i < n; i++)
        {
            if (temp.SequenceEqual(temp2))
                return true;
            var first = temp[0];
            temp.RemoveAt(0);
            temp.Add(first);
        }
        return false;
    }

    public static SortedDictionary<int, int> CalculateOccurrences(IReadOnlyList<int> input)
    {
        var occurrences = new SortedDictionary<int, int>();
        foreach (var occurrence in input)
            occurrences[occurrence] = occurrences.GetValueOrDefault(occurrence) + 1;
        return occurrences;
    }

    public static bool IsWinogradDeep(SortedDictionary<int, int> input, int size)
    {
        var counts = new HashSet<int>();
        foreach (var entry in input)
        {
            var count = entry.Value;
            if (count < 1 || count >= size)
                continue;
            if (!counts.Add(count))
                return false;
        }
        return counts.Count == size - 1;
    }

    public static bool IsErdosDeep(SortedDictionary<int, int> input)
    {
        var seen = new HashSet<int>();
        foreach (var pair in input)
        {
            if (!seen.Add(pair.Value))
                return false;
        }
        return true;
    }

    public static double CalculateRegressionEvenness(IReadOnlyList<int> rhythm, int totalTimeUnits)
    {
        var numNotes = rhythm.Count;
        if (numNotes == 0)
            return 0.0;
        var idealInterval = (double)totalTimeUnits / numNotes;
        var deviations = new double[numNotes];
        for (var i = 0; i < numNotes; i++)
            deviations[i] = Math.Abs(rhythm[i] - i * idealInterval);
        return deviations.Sum();
    }

    public static int CalculateRhythmicOddity(PositionVector input)
    {
        var k = (int)input.Count;
        var rhythmicOddity = 0;
        var mod = input.Mod;
        for (var i = 0; i < k; i++)
        {
            for (var j = i + 1; j < k; j++)
            {
                var dist1 = (input[j] - input[i] + mod) % mod;
                var dist2 = mod - dist1;
                if (dist1 == dist2)
                    rhythmicOddity++;
            }
        }
        return rhythmicOddity;
    }

    public static int ComputeTransitionComplexity(PositionVector input, int mod)
    {
        var onsetVec = MusicVectors.PositionsToOnset(input);
        if (onsetVec.Count == 0)
            return 0;
        var complexity = 0;
        for (var i = 1; i < onsetVec.Count; i++)
        {
            if (onsetVec[i] != onsetVec[i - 1])
                complexity++;
        }
        if (onsetVec[^1] != onsetVec[0])
            complexity++;
        return complexity;
    }

    public static double ComputeEntropy(PositionVector input)
    {
        var onsetVec = MusicVectors.PositionsToOnset(input);
        if (onsetVec.Count == 0)
            return 0.0;
        var frequency = new Dictionary<int, int>();
        foreach (var v in onsetVec.MutableData)
            frequency[v] = frequency.GetValueOrDefault(v) + 1;
        var entropy = 0.0;
        var n = onsetVec.Count;
        foreach (var pair in frequency)
        {
            var probability = (double)pair.Value / n;
            entropy -= probability * Math.Log2(probability);
        }
        return entropy;
    }

    public static int ComputeLongestSubsequence(PositionVector input)
    {
        var onsetVec = MusicVectors.PositionsToOnset(input);
        var data = onsetVec.MutableData;
        var n = data.Count;
        if (n == 0)
            return 0;
        var longest = 1;
        var currentRun = 1;
        for (var i = 1; i < n; i++)
        {
            if (data[i] == data[i - 1])
            {
                currentRun++;
                if (currentRun > longest)
                    longest = currentRun;
            }
            else
                currentRun = 1;
        }

        if (n >= 2 && data[0] == data[n - 1])
        {
            var wrapVal = data[0];
            var prefix = 0;
            while (prefix < n && data[prefix] == wrapVal)
                prefix++;
            var suffix = 0;
            while (suffix < n && data[n - 1 - suffix] == wrapVal)
                suffix++;
            if (prefix == n)
                longest = n;
            else
                longest = Math.Max(longest, prefix + suffix);
        }

        return longest;
    }

    public static void PrintDistances(PositionVector input, IReadOnlyList<int> distances)
    {
        var index = 0;
        var n = input.Count;
        for (var i = 0; i < n; i++)
        {
            for (var j = i + 1; j < n; j++)
                Console.WriteLine($"Distance between {input[i]} and {input[j]} is: {distances[index++]}");
        }
    }

    public static void PrintOccurrences(SortedDictionary<int, int> occurrences)
    {
        foreach (var pair in occurrences)
            Console.WriteLine($"Distance {pair.Key} appears {pair.Value} times");
    }

    public static void PrintDeepness(SortedDictionary<int, int> occurrences, int size)
    {
        var winogradDeep = IsWinogradDeep(occurrences, size);
        var erdosDeep = IsErdosDeep(occurrences);
        Console.WriteLine("The vector is " + (winogradDeep ? "" : "not ") + "Winograd-deep");
        Console.WriteLine("The vector is " + (erdosDeep ? "" : "not ") + "Erdos-deep");
    }

    public static List<SortedSet<int>> CalculateDistributionSpectra(PositionVector input)
    {
        var normalizedScale = input.MutableData.ToList();
        if (normalizedScale.Count == 0)
            return new List<SortedSet<int>>();
        var mod = input.Mod;
        var distributionSpectra = new List<SortedSet<int>>();
        for (var k = 0; k < normalizedScale.Count - 1; k++)
            distributionSpectra.Add(new SortedSet<int>());
        var sz = normalizedScale.Count;
        for (var i = 0; i < sz; i++)
        {
            for (var j = 1; j < sz; j++)
            {
                var genericInterval = j;
                var specificInterval = (normalizedScale[(i + j) % sz] - normalizedScale[i] + mod) % mod;
                distributionSpectra[genericInterval - 1].Add(specificInterval);
            }
        }
        return distributionSpectra;
    }

    public static List<int> CalculateSpectrumWidths(List<SortedSet<int>> spectra)
    {
        var widths = new List<int>();
        foreach (var spectrum in spectra)
        {
            if (spectrum.Count == 0)
                widths.Add(0);
            else
                widths.Add(spectrum.Max - spectrum.Min);
        }
        return widths;
    }

    public static double CalculateSpectrumVariation(IReadOnlyList<int> widths, int numberOfTones)
    {
        if (numberOfTones <= 0)
            return 0.0;
        return (double)widths.Sum() / numberOfTones;
    }

    public static List<int> FindRotationalSymmetryAxes(PositionVector scale)
    {
        var normalizedScale = scale.MutableData.ToList();
        var sortedNormalized = normalizedScale.OrderBy(x => x).ToList();
        var axes = new List<int>();
        var n = normalizedScale.Count;
        for (var interval = 1; interval < scale.Mod; interval++)
        {
            var transposedScale = new int[n];
            for (var i = 0; i < n; i++)
                transposedScale[i] = (normalizedScale[i] + interval) % scale.Mod;
            var sortedT = transposedScale.OrderBy(x => x).ToList();
            if (sortedT.SequenceEqual(sortedNormalized))
                axes.Add(interval);
        }
        return axes;
    }

    public static List<double> FindReflectiveSymmetryAxes(PositionVector scale)
    {
        var normalizedScale = scale.MutableData.ToList();
        var axes = new List<double>();
        var n = normalizedScale.Count;
        var modD = (double)scale.Mod;
        for (var axisDoubled = 0; axisDoubled < 2 * scale.Mod; axisDoubled++)
        {
            var axis = axisDoubled / 2.0;
            var isSymmetric = true;
            for (var i = 0; i < n; i++)
            {
                var reflected = 2 * axis - normalizedScale[i];
                var wrapped = reflected % modD;
                if (wrapped < 0)
                    wrapped += modD;
                var reflectedNote = (int)Math.Round(wrapped);
                reflectedNote = ((reflectedNote % scale.Mod) + scale.Mod) % scale.Mod;
                if (!normalizedScale.Contains(reflectedNote))
                {
                    isSymmetric = false;
                    break;
                }
            }
            if (isSymmetric)
                axes.Add(axis);
        }
        return axes;
    }

    public static bool IsPrime(int num)
    {
        if (num <= 1)
            return false;
        if (num == 2)
            return true;
        if (num % 2 == 0)
            return false;
        for (var i = 3; i <= (int)Math.Sqrt(num); i += 2)
        {
            if (num % i == 0)
                return false;
        }
        return true;
    }

    public static void ClassifyAksakRhythm(int mod)
    {
        if (IsPrime(mod))
            Console.WriteLine("The rhythm is authentic aksak");
        else if (mod % 2 != 0)
            Console.WriteLine("The rhythm is quasi-aksak");
        else
            Console.WriteLine("The rhythm is pseudo-aksak");
    }

    public static bool IsPalindrome(PositionVector scale) =>
        FindReflectiveSymmetryAxes(scale).Any(a => Math.Abs(a - 0.0) < 1e-9);

    public static bool IsChiral(PositionVector scale)
    {
        var normalizedScale = scale.MutableData.ToList();
        var sortedNormalized = normalizedScale.OrderBy(x => x).ToList();
        var mirroredScale = normalizedScale.Select(note => (scale.Mod - note) % scale.Mod).OrderBy(x => x).ToList();
        if (sortedNormalized.SequenceEqual(mirroredScale))
            return false;
        var n = normalizedScale.Count;
        for (var interval = 1; interval < scale.Mod; interval++)
        {
            var transposedMirrored = new int[n];
            for (var i = 0; i < n; i++)
                transposedMirrored[i] = (mirroredScale[i] + interval) % scale.Mod;
            var sortedT = transposedMirrored.OrderBy(x => x).ToList();
            if (sortedT.SequenceEqual(sortedNormalized))
                return false;
        }
        return true;
    }

    public static bool IsBalanced(PositionVector scale)
    {
        double xSum = 0, ySum = 0;
        var angleStep = 2 * Math.PI / scale.Mod;
        foreach (var note in scale.MutableData)
        {
            var angle = note * angleStep;
            xSum += Math.Cos(angle);
            ySum += Math.Sin(angle);
        }
        return Math.Abs(xSum) < 1e-6 && Math.Abs(ySum) < 1e-6;
    }

    public static List<int> Generate(int m, int k, int n, bool printSteps = false)
    {
        if (n <= 0)
            throw new ArgumentException("generate: modulus n must be positive");
        if (k < 0)
            throw new ArgumentException("generate: k must be non-negative");
        var sequence = new List<int>();
        for (var i = 0; i < k; i++)
        {
            var value = (i * m) % n;
            if (printSteps)
                Console.WriteLine($"{i} x {m} mod {n} = {value}");
            sequence.Add(value);
        }
        return sequence;
    }

    public static (bool Found, int Multiplier) IsGenerated(IReadOnlyList<int> input, int n)
    {
        var k = input.Count;
        var sortedIn = input.OrderBy(x => x).ToList();
        for (var mult = 1; mult < n; mult++)
        {
            var generatedSeq = Generate(mult, k, n);
            var sortedG = generatedSeq.OrderBy(x => x).ToList();
            if (sortedG.SequenceEqual(sortedIn))
                return (true, mult);
        }
        return (false, -1);
    }

    public static void PrintGenerators(IReadOnlyList<int> input, int mod)
    {
        var (found, m) = IsGenerated(input.ToList(), mod);
        if (found)
        {
            Console.WriteLine($"The vector is generated by multiples of m = {m} mod {mod}");
            Console.WriteLine("Generators:");
            Generate(m, input.Count, mod, true);
        }
        else
            Console.WriteLine($"The vector is not generated by multiples of any integer m under modulo {mod}");
    }

    public static void PrintDistributionSpectra(List<SortedSet<int>> spectra)
    {
        for (var i = 0; i < spectra.Count; i++)
            Console.WriteLine("<{0}> = {{{1}}}", i + 1, string.Join(",", spectra[i]));
    }

    public static void PrintSpectrumWidths(IReadOnlyList<int> widths)
    {
        for (var i = 0; i < widths.Count; i++)
            Console.WriteLine($"Width of <{i + 1}> = {widths[i]}");
    }

    public static void PrintSymmetryAxes(IReadOnlyList<int> axes, string symmetryType)
    {
        Console.Write(symmetryType + " symmetry axes: ");
        Console.WriteLine(string.Join(", ", axes));
    }

    public static void PrintSymmetryAxes(IReadOnlyList<double> axes, string symmetryType)
    {
        Console.Write(symmetryType + " symmetry axes: ");
        Console.WriteLine(string.Join(", ", axes));
    }

    public static void PrintAnalysis(PositionVector p)
    {
        var mod = p.Mod;
        var j = MusicVectors.PositionsToIntervals(p);
        var intervals = j.MutableData;
        var onsets = MusicVectors.PositionsToOnset(p);
        var distances = GeodesicDistances(p);
        var occurrences = CalculateOccurrences(distances);
        var spectra = CalculateDistributionSpectra(p);
        var widths = CalculateSpectrumWidths(spectra);
        var variation = CalculateSpectrumVariation(widths, (int)p.Count);

        Console.WriteLine("Analysis Results:");
        Console.WriteLine();
        Console.WriteLine("Modulo:");
        Console.WriteLine(mod);
        Console.WriteLine();
        Console.WriteLine("Positions:");
        Console.WriteLine(string.Join(" ", p.MutableData));
        Console.WriteLine();
        Console.WriteLine("Intervals:");
        Console.WriteLine(string.Join(" ", intervals));
        Console.WriteLine();
        Console.WriteLine("Onsets:");
        Console.WriteLine(string.Join(" ", onsets.MutableData));
        Console.WriteLine();
        Console.WriteLine("Distances:");
        PrintDistances(p, distances);
        Console.WriteLine();
        Console.WriteLine("Occurrences:");
        PrintOccurrences(occurrences);
        Console.WriteLine();
        Console.WriteLine("Deepness:");
        PrintDeepness(occurrences, p.MutableData.Count);
        Console.WriteLine();
        Console.WriteLine("Generators check:");
        PrintGenerators(p.MutableData, mod);
        Console.WriteLine();
        Console.WriteLine("Aksak Check:");
        ClassifyAksakRhythm(mod);
        Console.WriteLine();
        Console.WriteLine("Regression Evenness:");
        var rhythmList = p.MutableData.ToList();
        Console.WriteLine(CalculateRegressionEvenness(rhythmList, mod));
        Console.WriteLine();
        Console.WriteLine("Rhythmic Oddity:");
        Console.WriteLine(CalculateRhythmicOddity(p));
        Console.WriteLine();
        Console.WriteLine("Shannon Entropy:");
        Console.WriteLine(ComputeEntropy(p));
        Console.WriteLine();
        Console.WriteLine("Longest Subsequence:");
        Console.WriteLine(ComputeLongestSubsequence(p));
        Console.WriteLine();
        Console.WriteLine("Euclidean String Check:");
        var euclidean = IsEuclidean(p, mod);
        Console.WriteLine("The interval vector is " + (euclidean ? "" : "not ") + "an Euclidean string");
        Console.WriteLine();
        Console.WriteLine("Distribution Spectra:");
        PrintDistributionSpectra(spectra);
        Console.WriteLine();
        Console.WriteLine("Spectrum Widths:");
        PrintSpectrumWidths(widths);
        Console.WriteLine();
        Console.WriteLine("Spectrum Variation:");
        Console.WriteLine(variation);
        Console.WriteLine();
        Console.WriteLine("Symmetry Analysis:");
        var rot = FindRotationalSymmetryAxes(p);
        var refl = FindReflectiveSymmetryAxes(p);
        PrintSymmetryAxes(rot, "Rotational");
        PrintSymmetryAxes(refl, "Reflective");
        Console.WriteLine();
        Console.WriteLine("Transition complexity:");
        Console.WriteLine(ComputeTransitionComplexity(p, mod));
        Console.WriteLine();
        Console.WriteLine("Palindrome Check:");
        Console.WriteLine("The vector is " + (IsPalindrome(p) ? "" : "not ") + "palindrome");
        Console.WriteLine();
        Console.WriteLine("Chirality Check:");
        Console.WriteLine("The vector is " + (IsChiral(p) ? "" : "not ") + "chiral");
        Console.WriteLine();
        Console.WriteLine("Balance Check:");
        Console.WriteLine("The vector is " + (IsBalanced(p) ? "" : "not ") + "balanced");
        Console.WriteLine();
    }
}
